pkg/authorization/api/validation/validation.go

```diff
@@ -8,48 +8,126 @@ import (
 	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
 )
 
-// ValidatePolicyBinding tests required fields for a PolicyBinding.
-func ValidatePolicyBinding(policyBinding *authorizationapi.PolicyBinding) fielderrors.ValidationErrorList {
+func ValidatePolicyName(name string, prefix bool) (bool, string) {
+	if name != authorizationapi.PolicyName {
+		return false, "name must be " + authorizationapi.PolicyName
+	}
+
+	return true, ""
+}
+
+func ValidatePolicy(policy *authorizationapi.Policy) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
+	allErrs = append(allErrs, validation.ValidateObjectMeta(&policy.ObjectMeta, true, ValidatePolicyName).Prefix("metadata")...)
+
+	for roleKey, role := range policy.Roles {
+		if roleKey != role.Name {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("roles."+roleKey+".name", policy, "must be "+roleKey))
+		}
+
+		allErrs = append(allErrs, ValidateRole(&role).Prefix("roles."+roleKey)...)
```

Comment:
on a related note, is it an error to have two roles with the same name in a serialized Policy.Roles list? conversion doesn't seem to prevent that, and it makes surfacing errors keyed by rolename (rather than index) a bit odd